### PR TITLE
Remove `VerifiedOperation` :balloon: :wave: 

### PR DESCRIPTION
**Pull request by sandreae**
At some point during the _monster_ refactor at the end of NGI we realised we didn't need/want `VerifiedOperation` anymore, it could be better expressed with some generics and trait requirements.

I've done it here, and it felt goooooood.

## New Traits
- `WithOperationId` and `WithPublicKey` for structs which encapsulate published operations and any data that was signed and contains the public key of it's author
- these are used for some nice conversion traits like: `impl<T: AsOperation + WithOperationID + WithPublicKey> TryFrom<Vec<&T>> for Document {...}`

## Remove `VerifiedOperation`
- Most of the time this didn't even require using generics as simply passing tuples seemed more transparent and easy to work with in tests. Doubly true for lower level methods.

## Introduce `PublishedOperation` for `fixtures` and `MemoryStore` in `test_utils`
- This was required as we are working with a single generic type when retrieving them from the store, which still makes sense to me.

closes: #440 

## 📋 Checklist

- [x] Add tests that cover your changes
- [x] Add this PR to the _Unreleased_ section in `CHANGELOG.md`
- [x] Link this PR to any issues it closes
- [x] New files contain a SPDX license header
- [x] Check if descriptions and terminology match `handbook` content (and visa-versa)

### CODE REVIEW

File: p2panda-rs/src/document/document.rs [hunk 1]
```diff
@@ -100,8 +102,26 @@ impl Human for Document {
...
+
+    fn try_from(operations: &Vec<T>) -> Result<Self, Self::Error> {
+        let document_builder: DocumentBuilder = operations.into();
+        document_builder.build()
+    }
+}
+
 /// A struct for building [documents][`Document`] from a collection of [operations with
-/// metadata][`crate::operation::VerifiedOperation`].
+/// metadata][`crate::operation::PublishedOperation`].
```
Comment (sandreae) [inline p2panda-rs/src/document/document.rs:124]: Change this doc string test to not use `PublishedOperation`

File: p2panda-rs/src/document/document.rs [hunk 2]
```diff
@@ -111,13 +131,13 @@ impl Human for Document {
 /// # mod tests {
 /// # use rstest::rstest;
 /// # use p2panda_rs::document::DocumentBuilder;
-/// # use p2panda_rs::operation::VerifiedOperation;
+/// # use p2panda_rs::operation::PublishedOperation;
```
Comment (sandreae) [inline p2panda-rs/src/document/document.rs:134]: Here too.

File: p2panda-rs/src/document/document.rs [hunk 3]
```diff
@@ -265,32 +270,79 @@ impl DocumentBuilder {
     }
 }
 
+impl<T: AsOperation + WithOperationID + WithPublicKey> From<Vec<&T>> for DocumentBuilder {
+    fn from(operations: Vec<&T>) -> Self {
+        let operations = operations
```
Comment (sandreae) [inline p2panda-rs/src/document/document.rs:275]: Not very DRY in these conversion traits...

File: p2panda-rs/src/test_utils/db/types/operation.rs [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+use crate::document::DocumentViewId;
+use crate::identity::PublicKey;
+use crate::operation::traits::{AsOperation, WithOperationID, WithPublicKey};
+use crate::operation::{
+    Operation, OperationAction, OperationFields, OperationId, OperationVersion,
+};
+use crate::schema::SchemaId;
+
+/// On operation, it's id, and the public key of the keypair which signed it.
```
Comment (sandreae) [inline p2panda-rs/src/test_utils/db/types/operation.rs:11]: Typo "On"

File: p2panda-rs/src/document/document.rs (hunk 3 above)
Comment (adzialocha) [reply p2panda-rs/src/document/document.rs:251]: Yeah! I think it's alright, if you want you can use the `where` syntax if that makes it more readable?

I hope that Rust introduces a feature soon where one can define trait requirements similarily like `type`.

**Comment by sandreae** [comment]
@adzialocha ping pong!

I feel in the mood to kick off a couple of "New Year" code love PRs, how would you feel about merging this and dealing with any changes with issues later?

Think the codecov fail is an anomaly cos of an update on main.

**Comment by adzialocha** [comment]
> @adzialocha ping pong!
> 
> I feel in the mood to kick off a couple of "New Year" code love PRs, how would you feel about merging this and dealing with any changes with issues later?
> 
> Think the codecov fail is an anomaly cos of an update on main.

Don't want to merge it just like that but I'll give it a review soon ✊

**Comment by sandreae** [comment]
Also, if you'd rather not look at it till next year I can for sure work around it not being merged for any tidying I wanna do :+1:

File: p2panda-rs/src/operation/mod.rs [hunk 5]
```diff
@@ -117,7 +117,7 @@
 //!                                                 │
 //!                                                 ▼
 //!                                         ┌─────────────────┐
-//!                                         │VerifiedOperation│
+//!                                         │PublishedOperation│
```
Comment (adzialocha) [inline p2panda-rs/src/operation/mod.rs:120]: `PublishedOperation` is a type only used in `test_utils`, right? I'd not mention it here when its not part of the public API.

File: p2panda-rs/src/operation/traits.rs [hunk 6]
```diff
@@ -7,6 +7,22 @@ use crate::operation::plain::PlainFields;
...
+/// The method returns the public key of the keypair used to perform signing.
+pub trait WithPublicKey {
+    /// Returns the public key of the author of this entry or operation.
+    fn public_key(&self) -> &PublicKey;
+}
+
+/// Trait representing the id of an "operation-like" struct.
+///
+/// Returns an operation's id which is derived from the hash of the entry it was published with.
+pub trait WithOperationID {
```
Comment (adzialocha) [inline p2panda-rs/src/operation/traits.rs:21]: Maybe name it `WithOperationId` (with a small `d` at the end) for consistency (`SchemaId`, `OperationId`, etc.)

Comment (adzialocha) [reply p2panda-rs/src/operation/traits.rs:21]: .. and as written in another comment, `WithId` could also be quite cool

**Review by adzialocha — CHANGES_REQUESTED** [review]
This is awesome! Really good idea and nice additional conversion traits :sparkles: 

I mostly have some naming details, but nothing crucial.

File: p2panda-rs/src/document/document.rs [hunk 7]
```diff
@@ -100,59 +102,43 @@ impl Human for Document {
...
+
+    fn try_from(operations: Vec<&T>) -> Result<Self, Self::Error> {
+        let document_builder: DocumentBuilder = operations.into();
+        document_builder.build()
+    }
 }
 
+impl<T> TryFrom<&Vec<T>> for Document
+where
+    T: AsOperation + WithOperationID + WithPublicKey,
```
[the remark this answers is not in the document]
Comment (sandreae) [reply p2panda-rs/src/document/document.rs:119]: Yeh, that would be cool. Would you imagine it having a generic type parameter then so the id returned is actually generic?

File: p2panda-rs/src/operation/mod.rs (hunk 5 above)
Comment (sandreae) [reply p2panda-rs/src/operation/mod.rs:120]: Ah yes, this was an accident ;-p

**Review by adzialocha — APPROVED** [review]

**Pull request merged**